File: src/Algorithms/Sliding_Window_with_Frequency_Counter/3_Longest_Substring_Without_Repeating_Characters/3_Longest_Substring_Without_Repeating_Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int longestSubstring = 0;
        int[] freqCounter = new int[26];

        for (char ch : s.toCharArray()) {
            freqCounter[ch - 'a'] = 1;
        }

        int left = 0;
        for (int right = 0; right < s.length(); right++) {

            freqCounter[s.charAt(right) - 'a']--;

            while (freqCounter[s.charAt(right) - 'a'] < 0) {
                freqCounter[s.charAt(left) - 'a']++;
                left++;
            }

            longestSubstring = Math.max(longestSubstring, right - left + 1);
        }

        return longestSubstring;
    }
}

File: src/Algorithms/Sliding_Window_with_Frequency_Counter/3_Longest_Substring_Without_Repeating_Characters/3_Longest_Substring_Without_Repeating_Characters.js
/**
 * @param {string} s
 * @return {number}
 */

// create your own target string using the freqCounter

var lengthOfLongestSubstring = function(s) {
    const freqCounter = Array.from({ length: 26 }, () => 0);

    for (let i = 0; i < s.length; i += 1) {
        freqCounter[s.charCodeAt(i) - 'a'.charCodeAt(0)] = 1;
    }

    let maxLength = 0;

    let left = 0; 
    for (let right = 0; right < s.length; right += 1) {
        const curCharIndex = s.charCodeAt(right) - 'a'.charCodeAt(0);
        freqCounter[curCharIndex] -= 1;

        while (freqCounter[curCharIndex] < 0) {
            const leftCharIndex = s.charCodeAt(left) - 'a'.charCodeAt(0);
            freqCounter[leftCharIndex] += 1;
            left += 1;
        }

        maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
};